sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamZetaSqlCalcRel.java

```diff
@@ -0,0 +1,259 @@
+    RexNode condition;
+    if (programCondition == null) {
+      condition = null;
+    } else {
+      condition = program.expandLocalRef(programCondition);
+    }
+    // Currently this gets the selectivity based on Calcite's Selectivity Handler (RelMdSelectivity)
+    return mq.getSelectivity(child, condition);
+  }
+
+  @Override
+  public BeamCostModel beamComputeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+    NodeStats inputStat = BeamSqlRelUtils.getNodeStats(this.input, mq);
+    return BeamCostModel.FACTORY.makeCost(inputStat.getRowCount(), inputStat.getRate());
+  }
+
+  @Override
+  public boolean isInputSortRelAndLimitOnly() {
+    return (input instanceof BeamSortRel) && ((BeamSortRel) input).isLimitOnly();
+  }
+
+  /**
+   * {@code CalcFn} is the executor for a {@link BeamZetaSqlCalcRel} step. The implementation is
+   * based on the {@code ZetaSQL} expression evaluator.
+   */
+  private static class CalcFn extends DoFn<Row, Row> {
+    private final List<String> projects;
+    @Nullable private final String condition;
+    private final Schema inputSchema;
+    private final Schema outputSchema;
+    private transient List<PreparedExpression> projectExps;
+    @Nullable private transient PreparedExpression conditionExp;
+
+    CalcFn(
+        List<String> projects,
+        @Nullable String condition,
+        Schema inputSchema,
+        Schema outputSchema) {
+      Preconditions.checkArgument(projects.size() == outputSchema.getFieldCount());
+      this.projects = ImmutableList.copyOf(projects);
+      this.condition = condition;
+      this.inputSchema = inputSchema;
+      this.outputSchema = outputSchema;
+    }
+
+    @Setup
+    public void setup() {
+      AnalyzerOptions options = SqlAnalyzer.initAnalyzerOptions();
+      for (Field field : inputSchema.getFields()) {
+        options.addExpressionColumn(
+            sanitize(field.getName()), ZetaSqlUtils.beamFieldTypeToZetaSqlType(field.getType()));
+      }
+
+      projectExps = new ArrayList<>();
+      for (String project : projects) {
+        PreparedExpression projectExp = new PreparedExpression(sanitize(project));
+        projectExp.prepare(options);
+        projectExps.add(projectExp);
+      }
+      if (condition != null) {
+        conditionExp = new PreparedExpression(sanitize(condition));
+        conditionExp.prepare(options);
+      }
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      Map<String, Value> columns = new HashMap<>();
+      Row row = c.element();
+      for (Field field : inputSchema.getFields()) {
+        columns.put(
+            sanitize(field.getName()),
+            ZetaSqlUtils.javaObjectToZetaSqlValue(row.getValue(field.getName()), field.getType()));
+      }
+
+      // TODO[BEAM-8630]: support parameters in expression evaluation
+      // The map is empty because parameters in the query string have already been substituted.
+      Map<String, Value> params = new HashMap<>();
```

Comment:
nit: `Collections.<String, Value>emptyMap()` or something else like that.

Reply:
Thanks! Done.